File: design-pattern-jvm/app/src/main/java/design/pattern/java/patterns/behavioral/observer/NewsPaperSubscription.java
package design.pattern.java.patterns.behavioral.observer;

class NewsPaperSubscription {
  private NewsPaper observable;
  private NewsPaperReader observer;

  NewsPaperSubscription(NewsPaper observable, NewsPaperReader observer) {
    this.observable = observable;
    this.observer = observer;
  }

  void unsubscribe() {
    this.observable.readers.removeIf((reader) -> reader.equals(this.observer));
  }
}

File: design-pattern-go/src/patterns/behavioral/observer/NewsPaperSubscription.go
package observer

type newsPaperSubscription struct {
	observable *NewsPaper
	observer   *NewsPaperReader
}

func (n *newsPaperSubscription) Unsubscribe() {
	var newReaders = []*NewsPaperReader{}
	for _, reader := range n.observable.readers {
		if reader != n.observer {
			newReaders = append(newReaders, reader)
		}
	}
	n.observable.readers = newReaders
}
